examples/factory/src/lib.rs

```diff
@@ -0,0 +1,423 @@
+    pub enum ProductState {
+        /// The product has not yet been initialized by the factory.
+        Uninitialized,
+        /// The product has been initialised by the factory.
+        Initialized(Product),
+    }
+
+    /// Errors
+    #[derive(Serialize, Debug, PartialEq, Eq, Reject, SchemaType)]
+    pub enum ProductError {
+        /// Failed parsing the parameter.
+        #[from(ParseError)]
+        ParseParams,
+        /// The originator of the transaction does not match the owner of the
+        /// instance, and thus is not authorized to initialize the contract.
+        NotAuthorized,
+        /// The product was already initialized.
+        AlreadyInitialized,
+        /// [`initialize`] was not called by a contract.
+        NoFactoryContract,
+    }
+
+    /// Construct the product contract in the uninitialized state.
+    #[init(contract = "product")]
+    pub fn init(_ctx: &InitContext, _state_builder: &mut StateBuilder) -> InitResult<ProductState> {
+        Ok(ProductState::Uninitialized)
+    }
+
+    /// Initialize the product contract, supplying its index.
+    /// The following preconditions are checked:
+    ///
+    /// * The instance must not have already been initialized. Otherwise,
+    ///   [`ProductError::AlreadyInitialized`] is returned.
+    ///
+    /// * The account invoking the current transaction must be the same account
+    ///   that created the instance. Otherwise [`ProductError::NotAuthorized`]
+    ///   is returned.
+    ///
+    /// * The method must be called by another smart contract (the factory). If
+    ///   the caller is not a contract, [`ProductError::NoFactoryContract`] is
+    ///   returned.
+    ///
+    /// Note that beyond checking that the caller is a smart contract, this
+    /// method does not check further properties of the caller (e.g. that it is
+    /// any particular factory smart contract).
+    #[receive(
+        contract = "product",
+        name = "initialize",
+        parameter = "u64",
+        mutable,
+        return_value = "()",
+        error = "ProductError"
+    )]
+    pub fn initialize(
+        ctx: &ReceiveContext,
+        host: &mut Host<ProductState>,
+    ) -> Result<(), ProductError> {
+        let state = host.state_mut();
+        // Check that the contract has not already been initialized.
+        ensure_eq!(*state, ProductState::Uninitialized, ProductError::AlreadyInitialized);
+        // Check that the invoker is the owner of the contract.
+        ensure_eq!(ctx.invoker(), ctx.owner(), ProductError::NotAuthorized);
+        // The index is supplied as a parameter by the factory.
+        let index: u64 = ctx.parameter_cursor().get()?;
+        // This endpoint should only be called by another smart contract, namely the
+        // factory, which we record in the state.
+        let factory = match ctx.sender() {
+            Address::Contract(ca) => ca,
+            _ => Err(ProductError::NoFactoryContract)?,
```

Comment:
Suggestion: call it `OnlyFactoryContract/OnlyContract` as we do this in our other examples.

The `NoFactoryContract` is confusing because it could be interpreted as that this entry point cannot be invoked by a factory contact `NoFactoryContract`.

https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/auction/src/lib.rs#L133

Reply:
If the error should state what went wrong then the correct error description is:
`SenderIsAccountAddress`